src/algebra/module/graded_module.lean

```diff
@@ -0,0 +1,218 @@
+    (mul_smul (graded_monoid.mk ai ax) (graded_monoid.mk bi bx) (graded_monoid.mk ci cx)),
+end
+
+/-- The `module` derived from `gmodule A M`. -/
+instance gmodule.module [decidable_eq ι] [gsemiring A] [gmodule A M] :
+  module (⨁ i, A i) (⨁ i, M i) :=
+{ smul := (•),
+  one_smul := one_smul _ _,
+  mul_smul := mul_smul _ _,
+  smul_add := λ r, (gmodule.smul_add_monoid_hom A M r).map_add,
+  smul_zero := λ r, (gmodule.smul_add_monoid_hom A M r).map_zero,
+  add_smul := λ r s x, by simp only [gmodule.smul_def, map_add, add_monoid_hom.add_apply],
+  zero_smul := λ x, by simp only [gmodule.smul_def, map_zero, add_monoid_hom.zero_apply] }
+
+end
+
+end direct_sum
+
+end
+
+open_locale direct_sum big_operators
+
+variables {ι R A M σ σ' : Type*}
+variables [add_monoid ι] [comm_semiring R] [semiring A] [algebra R A]
+variables (𝓐 : ι → σ') [set_like σ' A]
+variables (𝓜 : ι → σ)
+
+namespace set_like
+
+include σ' A σ M
+
+instance gmul_action [add_monoid M] [distrib_mul_action A M]
+  [set_like σ M] [set_like.graded_monoid 𝓐] [set_like.has_graded_smul 𝓐 𝓜] :
+  graded_monoid.gmul_action (λ i, 𝓐 i) (λ i, 𝓜 i) :=
+{ one_smul := λ ⟨i, m⟩, sigma.subtype_ext (zero_add _) (one_smul _ _),
+  mul_smul := λ ⟨i, a⟩ ⟨j, a'⟩ ⟨k, b⟩, sigma.subtype_ext (add_assoc _ _ _) (mul_smul _ _ _),
+  ..set_like.ghas_smul 𝓐 𝓜 }
+
+instance gdistrib_mul_action [add_monoid M] [distrib_mul_action A M]
+  [set_like σ M] [add_submonoid_class σ M] [set_like.graded_monoid 𝓐]
+  [set_like.has_graded_smul 𝓐 𝓜] :
+  direct_sum.gdistrib_mul_action (λ i, 𝓐 i) (λ i, 𝓜 i) :=
+{ smul_add := λ i j a b c, subtype.ext $ smul_add _ _ _,
+  smul_zero := λ i j a, subtype.ext $ smul_zero _,
+  ..set_like.gmul_action 𝓐 𝓜 }
+
+variables [add_comm_monoid M] [module A M] [set_like σ M] [add_submonoid_class σ' A]
+  [add_submonoid_class σ M] [set_like.graded_monoid 𝓐] [set_like.has_graded_smul 𝓐 𝓜]
+
+/-- `[set_like.graded_monoid 𝓐] [set_like.has_graded_smul 𝓐 𝓜]` is the internal version of graded
+  module, the internal version can be translated into the external version `gmodule`. -/
+instance gmodule : direct_sum.gmodule (λ i, 𝓐 i) (λ i, 𝓜 i) :=
+{ smul := λ i j x y, ⟨(x : A) • (y : M), set_like.has_graded_smul.smul_mem x.2 y.2⟩,
+  add_smul := λ i j a a' b, subtype.ext $ add_smul _ _ _,
+  zero_smul := λ i j b, subtype.ext $ zero_smul _ _,
+  ..set_like.gdistrib_mul_action 𝓐 𝓜}
+
+end set_like
+
+namespace graded_module
+
+include σ' A σ M
+
+variables [add_comm_monoid M] [module A M] [set_like σ M] [add_submonoid_class σ' A]
+  [add_submonoid_class σ M] [set_like.graded_monoid 𝓐] [set_like.has_graded_smul 𝓐 𝓜]
+
+/--
+The smul multiplication of `A` on `⨁ i, 𝓜 i` from `(⨁ i, 𝓐 i) →+ (⨁ i, 𝓜 i) →+ ⨁ i, 𝓜 i`
+turns `⨁ i, 𝓜 i` into an `A`-module
+-/
+def is_module [decidable_eq ι] [graded_ring 𝓐] :
+  module A (⨁ i, 𝓜 i) :=
+{ smul := λ a b, direct_sum.decompose 𝓐 a • b,
+  ..module.comp_hom _ $
+    (direct_sum.decompose_ring_equiv 𝓐 : A ≃+* ⨁ i, 𝓐 i).to_ring_hom }
```

Comment:
This one worries me a bit; does `⨁ i, 𝓜 i` not already inherit a different action by `A`?